semistruct.go

```diff
@@ -0,0 +1,266 @@
+package semistruct
+
+import "strconv"
+import p "github.com/andyleap/parser"
+
+type Semistruct_log struct {
+	Priority int64
+	Tags     []string
+	Attrs    map[string]string
+}
+
+type kv_pair struct {
+	key string
+	val string
+}
+
+// Parse a semistructured log line into the semistruct_log struct
+// type.
+func ParseSemistruct() *p.Grammar {
+	o := p.And(
+		// Parse opening sentinel "!<"
+		OpenSentinel(),
+		SkipSpace(),
+
+		// Parse our log-level priority
+		p.Tag("priority", PriorityInt()),
+		SkipSpace(),
+
+		// Parse our tag list "[tag1:tag2:tag3]"
+		p.Tag("tags", Tags()),
+		SkipSpace(),
+
+		// Parse our attribute set "{ key=val key2=val2 }"
+		p.Tag("attrs", Attrs()),
+		SkipSpace(),
+
+		// Parse the ending sentinel ">!"
+		EndSentinel(),
+	)
+
+	o.Node(func(m p.Match) (p.Match, error) {
+		pr := p.GetTag(m, "priority").(int64)
+		tg := p.GetTag(m, "tags").([]string)
+		at := p.GetTag(m, "attrs").(map[string]string)
+
+		return Semistruct_log{pr, tg, at}, nil
+	})
+
+	return o
+}
+
+// TODO: Make sure all fields in the schema that are optional are
+// handled in here with good mempty-like defaults (or is there an
+// option type analog in Go?).
+//
+// TODO: Rename some of these parser combinators.
+//
+// TODO: Refactor the repetitive Mult calls all over the place if I
+// can, there's a lot of repetition I'd like to get rid of on my first
+// pass.
+//
+// TODO: handle failure better - if the string can't parse, what do we
+// do?
+//
+// TODO: are we consuming to the end of the line???
+
+// Parse the log level priority indicator; this can only be an integer
+// from 0 to 9
+func PriorityInt() *p.Grammar {
+	o := p.Set("0-9")
+	o.Node(func(m p.Match) (p.Match, error) {
+		v, err := strconv.ParseInt(p.String(m), 10, 64)
+		if err != nil {
+			return nil, err
+		}
+		return v, nil
+	})
+
+	return o
+}
+
+// Parse zero or more characters that are alpha-numeric (note, \w
+// includes the underscore).
+func AlphaNum() *p.Grammar {
+	return p.Mult(
+		0, 0, p.Set("\\w\\-"),
```

Comment:
can these p.Set functions be factored into vars, since they seem to be repeatedly creating new ones.